db/queries/base.py

```diff
@@ -1,203 +1,145 @@
-        by this query.
+        Sequence of SQLAlchemy columns representing the output columns of the
+        relation described by this query.
         """
         return tuple(
-            MathesarColumn.from_column(sa_col, engine=engine)
+            MathesarColumn.from_column(sa_col, engine=self.engine)
             for sa_col
-            in self.sa_relation.columns
+            in self.transformed_relation.columns
         )
 
-    @cached_property
-    def sa_relation(self):
+    @property
+    def transformed_relation(self):
         """
-        A query describes a relation. This property is the result of parsing a query into a
-        relation.
+        A query describes a relation. This property is the result of parsing a
+        query into a relation.
         """
-        initial_relation = _get_initial_relation(self)
         transformations = self.transformations
         if transformations:
             transformed = apply_transformations(
-                initial_relation,
+                self.initial_relation,
                 transformations,
             )
             return transformed
         else:
-            return initial_relation
+            return self.initial_relation
+
+    @property
+    def initial_relation(self):
+        metadata = MetaData()
+        base_table = reflect_table_from_oid(
+            self.base_table_oid, self.engine, metadata=metadata
+        )
+        from_clause = base_table
+        # We cache this to avoid copies of the same join path to a given table
+        jp_path_alias_map = {(): base_table}
+
+        def _get_table(oid):
+            """
+            We use the function-scoped metadata so all involved tables are aware
+            of each other.
+            """
+            return reflect_table_from_oid(oid, self.engine, metadata=metadata)
+
+        def _get_column_name(oid, attnum):
+            return get_column_name_from_attnum(oid, attnum, self.engine)
+
+        def _process_initial_column(col):
+            nonlocal from_clause
+            col_name = _get_column_name(col.reloid, col.attnum)
+            # Make the path hashable so it can be a dict key
+            jp_path = _guarantee_jp_path_tuples(col.jp_path)
+            right = base_table
+
+            for i, jp in enumerate(jp_path):
+                left = jp_path_alias_map[jp_path[:i]]
+                right = _get_table(jp[1][0]).alias()
+                jp_path_alias_map[jp_path[:i + 1]] = right
+                left_col = left.columns[_get_column_name(jp[0][0], jp[0][1])]
+                right_col = right.columns[_get_column_name(jp[1][0], jp[1][1])]
+                from_clause = from_clause.join(
+                    right, onclause=left_col == right_col, isouter=True,
+                )
+
+            return right.columns[col_name].label(col.alias)
+
+        stmt = select(
+            [_process_initial_column(col) for col in self.initial_columns]
```

Comment:
The inner function can't return the join statement. It has to build on the same global cache. Otherwise you end up with duplicates (or you need another piece of logic to remove them). This is why I went with the kind of "hybrid approach" with the side effect for the join, and the return for the column. I suppose I could make both side effects (i.e., also build the select column list via a side effect), but I'd rather minimize mutating things all over the place.

Reply:
A comment explaining that in the codebase might be helpful for the future.